src/storage/v2/durability/snapshot.cpp

```diff
@@ -127,6 +139,230 @@ struct BatchInfo {
+ public:
+  using task_t = std::function<void()>;
+  SafeTaskQueue() = default;
+
+  // Add a task to the queue.
+  void AddTask(std::function<void()> task) {
+    std::lock_guard<std::mutex> lock(mtx_);
+    tasks_.push(std::move(task));
+  }
+
+  std::optional<task_t> PopTask() {
+    std::lock_guard<std::mutex> lock(mtx_);
+    if (tasks_.empty()) {
+      return std::nullopt;
+    }
+    auto task = std::move(tasks_.front());
+    tasks_.pop();
+    return std::move(task);
+  }
+
+  auto size() const {
+    std::lock_guard<std::mutex> lock(mtx_);
+    return tasks_.size();
+  }
+
+ private:
+  mutable std::mutex mtx_;
+  std::queue<task_t> tasks_;
+};
+
+void WaitAndCombine(std::vector<std::pair<SnapshotBatchRes, std::promise<bool>>> &edge_res,
+                    std::vector<std::pair<SnapshotBatchRes, std::promise<bool>>> &vertex_res, uint64_t &offset_edges,
+                    uint64_t &offset_vertices, Encoder &snapshot, uint64_t &edges_count, uint64_t &vertices_count,
+                    std::vector<BatchInfo> &edge_batch_infos, std::vector<BatchInfo> &vertex_batch_infos,
+                    std::unordered_set<uint64_t> &used_ids) {
+  int i = 0;
+  // TODO All of this combination logic is only for multi-threaded; Make it into a fucntion and reuse
+  // Start combining files as soon as they come in
+  // NOTE: They have to be combined in order
+  if (!edge_res.empty()) offset_edges = snapshot.GetPosition();  // Global edge offset; 0 -> edges without properties
+  for (auto &[res, promise] : edge_res) {
+    promise.get_future().wait();  // Wait for incoming result
+    spdlog::trace("Handling snapshot edge part {}, size {}, count {}...", i++, res.snapshot_size, res.count);
+    if (res.snapshot_size > 0) {
+      edges_count += res.count;
+      used_ids.merge(std::move(res.used_ids));
+      const auto current_offset = snapshot.GetPosition();
+      // Update batch positions
+      for (auto &[offset, _] : res.batch_info) {
+        offset += current_offset;
+      }
+      // TODO Regenerate batches (currently could be less then optimal)
+      edge_batch_infos.insert(edge_batch_infos.end(), res.batch_info.begin(), res.batch_info.end());
+      // Append the edge part to the snapshot
+      int edges_fd = open(res.snapshot_path.string().c_str(), O_RDONLY);
+      if (edges_fd == -1) {
+        throw RecoveryFailure("Couldn't open edge snapshot part!");
+      }
+      utils::OnScopeExit cleanup{[&] {
+        close(edges_fd);
+        std::error_code ec;
+        std::filesystem::remove(res.snapshot_path, ec);
+        if (ec) spdlog::warn("Couldn't remove temporary edge snapshot part {}: {}", res.snapshot_path, ec.message());
+      }};
+      // Use sendfile for efficient copying (zero-copy)
+      off_t offset = 0;
+      const auto size = res.snapshot_size - offset;
+      ssize_t bytes_sent = sendfile(snapshot.native_handle(), edges_fd, &offset, size);
+      if (bytes_sent == -1 || bytes_sent != size) {
```

Comment:
Maybe makes sense to abstract this part into some utils method or something similar?